lib/rust/ensogl/core/src/animation/frp/animation/overshoot.rs

```diff
@@ -0,0 +1,134 @@
+//! Defines an animation driver for value with soft bounds.
+//!
+//! This animation is used to animate values with soft bounds. Depending on the input method, the
+//! target value might be allowed to exceed the bounds, but after very short time it will smoothly
+//! return into the bounds. This is useful for animating scrollbars, where scrolling past the end of
+//! the content should be possible for a short time, but the scrollbar should quickly return to the
+//! bounds. The sensitivity for setting the value past the bounds is reduced and value is limited to
+//! prevent the overshoot from being too large or too jumpy.
+
+use crate::prelude::*;
+
+use crate::animation::delayed::DelayedAnimation;
+use crate::animation::linear_interpolation;
+use crate::Animation;
+
+use enso_frp as frp;
+
+
+
+// ==================
+// === Constants  ===
+// ==================
+
+/// Determines how much of out-of-bounds portion of value change is actually applied to the
+/// animation target value. Larger values will make the overshoot more noticeable.
+const OUT_OF_BOUNDS_SENSITIVITY: f32 = 0.3;
+
+
+// ===========
+// === Frp ===
+// ===========
+
+crate::define_endpoints! {
+    Input {
+        /// Apply relative change to target value. Allow the value to overshoot.
+        soft_change_by (f32),
+        /// Apply relative change to target value. Remove any overshoot.
+        hard_change_by (f32),
+        /// Apply absolute change to target value. Allow the value to overshoot.
+        soft_change_to (f32),
+        /// Apply absolute change to target value, Remove any overshoot.
+        hard_change_to (f32),
+        /// Set the minimum bound of the target value. Values smaller than set bound will be
+        /// considered as overshoot and will bounce back.
+        set_min_bound (f32),
+        /// Set the maximum bound of the target value. Values larger than set bound will be
+        /// considered as overshoot will bounce back.
+        set_max_bound (f32),
+        /// Set the absolute maximum overshoot amplitude that the target value can have. If the
+        /// value overshoots more than set limit, it will be immediately clamped to that limit.
+        set_overshoot_limit (f32),
+        /// Set the time interval from when soft value change is applied to when bounce back
+        /// animation starts.
+        set_bounce_delay (f32),
+        /// Skip the animation and set the target value to the current value.
+        skip (),
+    }
+    Output {
+        /// The current value of the animation.
+        value (f32),
+        /// Current target value of the animation.
+        target (f32),
+    }
+}
+
+
+
+// ==========================
+// === OvershootAnimation ===
+// ==========================
+
+/// An [`Animation`] wrapper that allows the value to be smoothly manipulated within set bounds,
+/// overshoot set bounds and bounce back.
+#[derive(Clone, CloneRef, Debug, Shrinkwrap)]
```

Comment:
This does not explain how to use this structure and how it works. I feel that this is a very cool struct, it should have really good docs.

Reply:
Added more detailed documentation how to use the provided API and what's the animation behavior. There is also an overview already in the top of the module.